Enter edit mode if updating roles fails during user creation

rak-phillip opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
This emits  an `update:mode` event that sets to mode to "edit" and refreshes all resource-related variables when the `updateRoles()` method fails during user creation. This provides an opportunity for users to correct mistakes while still staying in the "Create User" context.

Fixes #13539 
<!-- Define findings related to the feature or bug issue. -->

### Occurred changes and/or fixed issues
<!-- Include information of the changes, including collateral areas which have been affected by this PR as requirement or for convenience. -->

- Switch mode to "edit" if the `updateRoles()` method fails during user creation
- Add a user readable error message if an escalation error is raised

### Technical notes summary
<!-- Outline technical changes which may pass unobserved or may help to understand the process of solving the issue -->

An alternate approach that explored navigating to an edit route was explored, but this required re-rendering the entire form and would eliminate the error message; additional state to persist the error messages would be required and I found that requirement to be more complex than the approach taken here.

### Areas or cases that should be tested
<!-- Areas that should be tested can include Airgap checks, Rancher upgrades, K8s upgrade, etc. -->
<!-- Which browser did you use for local testing? The reviewer should test with a different browser. -->
<!-- Add missing steps or rewrite them if have been missed or to complement existing information. This should define a clear way to reproduce it and not an approximation. -->

User Creation: See details in the associated issue.

### Areas which could experience regressions
<!-- Create a detailed list of areas to be analyzed which may be affected by the changes, which would require a prior research to avoid regressions. -->

User Creation: Ensuring that the form still functions properly after an error is raised during user creation.

### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->

![image](https://github.com/user-attachments/assets/57b4c9c9-85fe-45cf-b32f-a1af02fd706d)

### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone <!-- The milestone should automatically be assigned if the linked issue has one, but does not always happen (incorrectly linked, issue has no milestone, etc) -->
- [x] The PR template has been filled out
- [x] The PR has been self reviewed <!-- There are no TODOs, no incorrect files in the PR, all the required files are there, no commented out code, etc-->
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

rak-phillip commented on shell/assets/translations/en-us.yaml:
```diff
@@ -5134,6 +5134,8 @@ rbac:
       clustertemplaterevisions-create:
         label: Create RKE Template Revisions
+  errors:
+    escalation: You do not have sufficient permissions to grant Global Permissions above your own.
```
I'm open to suggestions for improving this error message. This is what the original error looks like

```
admission webhook "rancher.cattle.io.globalrolebindings.management.cattle.io" denied the request: errors due to escalation: user "u-q46zm" (groups=["system:authenticated" "system:cattle:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:["*"], Resources:["*"], Verbs:["*"]}
{NonResourceURLs:["*"], Verbs:["*"]}
```

eva-vashkevich replied:
I feel like you can just say:
"You cannot grant permissions above your own. Please verify your permissions." or something along those lines. There are no permissions that would be sufficient to grant someone permissions above your own

rak-phillip replied:
> There are no permissions that would be sufficient to grant someone permissions above your own

Right, that's logical in hindsight.. Thanks for the feedback 🙌

rak-phillip commented on shell/components/ResourceDetail/index.vue:
```diff
@@ -414,6 +403,50 @@ export default {
+     * @param {string} payload.resource - The resource string to use for
+     * initialization.
+     */
+    setMode({ mode, user, resource }) {
+      this.mode = mode;
+      this.configureResource(user, resource);
+    }
```
I'm not 100% satisfied with this approach, but it satisfies the constraint of allowing a user to continue making modifications if roles fail to assign during user creation for any reason. I'm open to alternative approaches..

rak-phillip commented:
@yonasberhe23 can you provide feedback on the e2e test for this PR?

yonasberhe23 commented on cypress/e2e/tests/pages/users-and-auth/users.spec.ts:
```diff
@@ -496,4 +498,62 @@ describe('Users', { tags: ['@usersAndAuths', '@adminUser'] }, () => {
+
+    it('User creation should complete after admin user fails to create for Standard user with Manage Users Role', () => {
+      // create standard user
+      usersPo.goTo();
```
I think its good practice to add `usersPo.waitForPage();` after navigation

yonasberhe23 commented on cypress/e2e/tests/pages/users-and-auth/users.spec.ts:
```diff
@@ -496,4 +498,62 @@ describe('Users', { tags: ['@usersAndAuths', '@adminUser'] }, () => {
+
+      // navigate to the users page and attempt to create an admin user
+      HomePagePo.goToAndWaitForGet();
+      usersPo.goTo();
```
add `usersPo.waitForPage();`

yonasberhe23 reviewed (changes requested):
Just two comments for the e2e test. Otherwise looks great!

codyrancher reviewed (approved):
Verified it works. 

I did notice that upon error it will change the global permissions back to a standard user. I think it could cause some confusion if the user goes back to check why the error was received.

rak-phillip commented:
> Verified it works.
>
> I did notice that upon error it will change the global permissions back to a standard user. I think it could cause some confusion if the user goes back to check why the error was received.

@codyrancher I fixed this behavior in 2fffa89ecda28885b83ef100471f9d9ef1961ce4. It turns out that the `userId` watcher was overriding values when we switched to the "edit" mode. I'm opting to override the watcher in this scenario to prevent the update.

rak-phillip commented:
All e2e tests pass, looking for a final review.
